libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright © 2017 Endless Mobile, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Sam Spilsbury <sam@endlessm.com>
+ */
+
+#include "enums.h"
+#include "types.h"
+
+#include <string.h>
+#include <errno.h>
+
+#include <flatpak.h>
+#include <json-glib/json-glib.h>
+#include <ostree.h>
+
+#include "flatpak.h"
+#include "util.h"
+
+static FlatpakRemoteRefAction *
+flatpak_remote_ref_action_new (EosUpdaterUtilFlatpakRemoteRefActionType  type,
+                               FlatpakRemoteRef                         *ref,
+                               gint32                                    serial)
+{
+  FlatpakRemoteRefAction *action = g_slice_new0 (FlatpakRemoteRefAction);
+
+  action->ref_cnt = 1;
+  action->type = type;
+  action->ref = g_object_ref (ref);
+  action->serial = serial;
+
+  return action;
+}
+
+FlatpakRemoteRefAction *
+flatpak_remote_ref_action_ref (FlatpakRemoteRefAction *action)
+{
+  ++action->ref_cnt;
+  return action;
+}
+
+void
+flatpak_remote_ref_action_unref (FlatpakRemoteRefAction *action)
+{
+  if (--action->ref_cnt != 0)
+    return;
+
+  g_object_unref (action->ref);
+
+  g_slice_free (FlatpakRemoteRefAction, action);
+}
+
+static gboolean
+flatpak_remote_ref_action_type_parse (const gchar                               *action,
+                                      EosUpdaterUtilFlatpakRemoteRefActionType  *out_action_type,
+                                      GError                                   **error)
+{
+  GEnumClass *enum_class = g_type_class_ref (EOS_TYPE_UPDATER_UTIL_FLATPAK_REMOTE_REF_ACTION_TYPE);
+  GEnumValue *enum_value = g_enum_get_value_by_nick (enum_class, action);
```

Comment:
I’d prefer using explicit hard-coded strings here, rather than the nicknames from the enum definition. It’s more explicit, and less likely to magically break if we refactor the code later and change the enum.

Reply:
@cosimoc Given that you were the one who recommended doing it this way, did you have any thoughts given the above ^ ?